Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/DocumentProvider.kt

```diff
@@ -0,0 +1,242 @@
+    val file = documentIdToPath(documentId)
+    val folder = documentIdToPath(parentDocumentId)
+    return file.relativeToOrNull(folder) != null
+  }
+
+  private fun appendDocument(file: File, cursor: MatrixCursor) {
+    var flags = 0
+    if (file.isDirectory && file.canWrite()) {
+      flags = DocumentsContract.Document.FLAG_DIR_SUPPORTS_CREATE
+    } else if (file.canWrite()) {
+      flags = DocumentsContract.Document.FLAG_SUPPORTS_WRITE
+      flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_DELETE
+      flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_REMOVE
+      flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_MOVE
+      flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_COPY
+      flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_RENAME
+    }
+
+    val name = if (file == rootDirectory) { context!!.getString(R.string.app_name) } else { file.name }
+    cursor.newRow().apply {
+      add(DocumentsContract.Document.COLUMN_DOCUMENT_ID, pathToDocumentId(file))
+      add(DocumentsContract.Document.COLUMN_MIME_TYPE, getTypeForFile(file))
+      add(DocumentsContract.Document.COLUMN_DISPLAY_NAME, name)
+      add(DocumentsContract.Document.COLUMN_LAST_MODIFIED, file.lastModified())
+      add(DocumentsContract.Document.COLUMN_FLAGS, flags)
+      add(DocumentsContract.Document.COLUMN_SIZE, file.length())
+      if (file == rootDirectory) {
+        add(DocumentsContract.Document.COLUMN_ICON, R.drawable.ic_dolphin)
+      }
+    }
+  }
+
+  private fun pathToDocumentId(path: File): String {
+    return "$ROOT_ID/${path.toRelativeString(rootDirectory!!)}"
+  }
+
+  private fun documentIdToPath(documentId: String): File {
+    val file = File(rootDirectory, documentId.substring("$ROOT_ID/".lastIndex))
+    if (!file.exists()) {
+      throw FileNotFoundException("File $documentId does not exist.")
+    }
+    return file
+  }
+
+  private fun getTypeForFile(file: File) : String {
+    return if (file.isDirectory)
+      DocumentsContract.Document.MIME_TYPE_DIR
+    else
+      MimeTypeMap.getSingleton().getMimeTypeFromExtension(file.extension)
+        ?: "application/octet-stream"
+  }
+
+  private fun findFileNameForNewFile(file: File) : File {
+    if (!shouldReplaceFile(file)) {
+      return file
+    }
+
+    var file = file
+    var i = 1
+    while (file.exists()) {
+      val pathWithoutExtension = file.absolutePath.substringBeforeLast('.')
+      val extension = file.absolutePath.substringAfterLast('.')
+      file = File("$pathWithoutExtension.$i.$extension")
+      i++
+    }
+    return file
+  }
+
+  private fun shouldReplaceFile(file: File) : Boolean {
+    val path = file.absolutePath
+    return !path.contains(WII_SAVE_DIR)
+      && !path.contains(GC_SAVE_DIR)
+      && !path.contains(SAVE_STATE_DIR)
+  }
+
+  private fun copyRecursively(src: File, dst: File): File {
+    val dst = findFileNameForNewFile(dst)
```

Comment:
What do you think about having names shadowed like this? I understand that it's a little cleaner but I think I prefer different names for clarity.

Reply:
I'm personally fine with shadowing (otherwise I wouldn't done it.).

The problem here is that there isn't really any much more descriptive name. I'd probably use something like `actualDst` instead of shadowing.

It's even worse at a different point where I shadow a name just to get a mutable variable.

Reply:
Fair, I won't make a big deal out of it

Reply:
I changed it. It's nice to get rid of the IDE notices.